### PR TITLE
Create Invoice Rebilling Service

### PR DESCRIPTION
**Pull request by StuAA78**
https://trello.com/c/qkWbpJlf/1962-m-create-re-billing-transactions

This change introduces the `InvoiceRebillingService` which accepts an invoice and will copy its licences and transactions to the supplied cancel invoice (with transaction types swapped so that credits become debits and vice-versa) and rebill invoice.

During development we found that `InvoiceModel` needed to be amended to accommodate `rebilledType` -- this change is included in this PR.

### CODE REVIEW

File: test/services/invoice_rebilling_create_transaction.service.test.js [hunk 1]
```diff
@@ -132,4 +132,14 @@ describe('Invoice Rebilling Create Transaction service', () => {
       })
     })
   })
+
+  async function addRebillInvoice (billRunId, customerReference, financialYear, rebilledInvoiceId, rebilledType) {
```
Comment (Cruikshanks) [inline test/services/invoice_rebilling_create_transaction.service.test.js:136]: How would you feel about putting this actually on the `InvoiceHelper` instead? It's a personal preference but if I can within tests I like to try and segregate logic/code into support methods and keep the test files focused on tests. 🤷 🤔

File: test/services/invoice_rebilling.service.test.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+  async function addRebillInvoice (billRunId, customerReference, financialYear, rebilledInvoiceId, rebilledType) {
+    return InvoiceHelper.addInvoice(
+      billRunId,
+      customerReference,
+      financialYear,
+      0, 0, 0, 0, 0, 0, 0, 0,
+      rebilledInvoiceId,
+      rebilledType)
+  }
```
Comment (Cruikshanks) [inline test/services/invoice_rebilling.service.test.js:145]: Similar to my previous comment I think these might be better moved to the helpers (though which ones will probably involve a discussion!).

Something like `createPopulatedLicence()` especially. I think it and some other test setups I've seen indicate we have a real need to create populated bill runs for testing.

File: test/services/invoice_rebilling.service.test.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    })
+
+    describe('cancel licences', () => {
+      let licences
+
+      beforeEach(async () => {
+        licences = await LicenceModel.query().where('invoiceId', cancelInvoice.id)
+      })
+
+      it('are created for each licence of the invoice', async () => {
```
Comment (Cruikshanks) [inline test/services/invoice_rebilling.service.test.js:63]:
```suggestion
      it('are created for each licence of the original invoice', async () => {
```

File: test/services/invoice_rebilling.service.test.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    })
+
+    describe('rebill licences', () => {
+      let licences
+
+      beforeEach(async () => {
+        licences = await LicenceModel.query().where('invoiceId', rebillInvoice.id)
+      })
+
+      it('are created for each licence of the invoice', async () => {
```
Comment (Cruikshanks) [inline test/services/invoice_rebilling.service.test.js:90]:
```suggestion
      it('are created for each licence of the original invoice', async () => {
```

**Review by Cruikshanks — COMMENTED** [review]
I think we just need to have a conversation about test helpers and whether you also have seen a growing need for helpers that can create properly populated bill runs, invoices and licences.

There is a good chance once we've had a discussion this will go as is and we will then return to it as part of housekeeping. But I think I'd like to have the discussion first in case anything comes up.

File: test/services/invoice_rebilling.service.test.js (hunk 2 above)
Comment (StuAA78) [reply test/services/invoice_rebilling.service.test.js:145]: As discussed we're leaving this for now due to circular referencing issues and we'll revisit test helpers later.

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**